### PR TITLE
[FLINK-18157][runtime] Jobstore size check compares against offHeapMemory

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Setting `jobmanager.memory.off-heap.size` to 0 results in this confusing error:`The configured or derived JVM heap memory size (jobstore.cache-size: 0 bytes) is less than the configured or default size of the job store cache (jobmanager.memory.heap.size: 50.000mb (52428800 bytes))`.  This log is confused for users, and should correct the content of this warn log. Method `verifyJobStoreCacheSize` in `JobManagerFlinkMemoryUtils` should correct validation result of jobstore cache size with jvm heap size.*

## Brief change log

  - *Correct warn log of method `verifyJobStoreCacheSize` in `JobManagerFlinkMemoryUtils`.*

## Verifying this change

  - *Add test method `testLogVerifyJvmHeapSizeAndJobStoreCacheSize` in `JobManagerProcessUtilsTest` to verfiy for JVM-Heap memory configuration and the case that JVM-Heap memory is less than default jobstore cache size.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
@azagrebin Could you please review this pull request for fixing jobstore cache size check bug？

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java [hunk 1]
```diff
@@ -88,16 +88,16 @@ private static void verifyJvmHeapSize(MemorySize jvmHeapSize) {
 		}
 	}
 
-	private static void verifyJobStoreCacheSize(Configuration config, MemorySize jvmHeapSize) {
+	private static void verifyJobStoreCacheSize(Configuration config, MemorySize offHeapMemorySize) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java:91]: This method is correct as the job cache is part of the JVM heap, not off-heap.
The problem is that `verifyJobStoreCacheSize` is called with `offHeapMemory` in `JobManagerFlinkMemoryUtils#createJobManagerFlinkMemory`:
`verifyJobStoreCacheSize(config, offHeapMemory);`
which should be:
`verifyJobStoreCacheSize(config, jvmHeap);` there.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java [hunk 2]
```diff
@@ -69,6 +69,20 @@ public void testConfigOffHeapMemory() {
 			jobManagerProcessSpec -> assertThat(jobManagerProcessSpec.getJvmDirectMemorySize(), is(offHeapMemory)));
 	}
 
+	@Test
+	public void testConfigJvmHeapAndOffHeapMemory() {
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java:73]: This test tests 2 things: heap and off-heap settings.
Also, I think it is redundant as it is duplicated with `testConfigJvmHeapMemory` and `testConfigOffHeapMemory`

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java (hunk 1 above)
Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java]: > This method is correct as the job cache is part of the JVM heap, not off-heap.
> The problem is that `verifyJobStoreCacheSize` is called with `offHeapMemory` in `JobManagerFlinkMemoryUtils#createJobManagerFlinkMemory`:
> `verifyJobStoreCacheSize(config, offHeapMemory);`
> which should be:
> `verifyJobStoreCacheSize(config, jvmHeap);` there.

@azagrebin Sorry for previous misunderstanding for the problem. And I have updated this method.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java (hunk 2 above)
Comment (SteNicholas) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java]: @azagrebin I previously want to add test for the case that the off-heap size is less than jobstore cache size. I have removed this test method.

**Comment by SteNicholas** [comment]
> Thanks for opening the PR @SteNicholas
> I have left comments to resolve.
> You could also look into `TestLoggerResource` usages to test the log message.

I would add test method to test the log message for this case immediately. Please check after I add this case.

**Comment by SteNicholas** [comment]
@azagrebin I have already modified the test method with `TestLoggerResource `. Please review again, thanks.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java [hunk 3]
```diff
@@ -42,19 +49,37 @@
...
-		MemorySize jvmHeapSize = MemorySize.parse("50m");
+		MemorySize jvmHeapMemory = MemorySize.parse("50m");
 
 		Configuration conf = new Configuration();
-		conf.set(JobManagerOptions.JVM_HEAP_MEMORY, jvmHeapSize);
+		conf.set(JobManagerOptions.JVM_HEAP_MEMORY, jvmHeapMemory);
+
+		JobManagerProcessSpec jobManagerProcessSpec = JobManagerProcessUtils.processSpecFromConfig(conf);
+		assertThat(jobManagerProcessSpec.getJvmHeapMemorySize(), is(jvmHeapMemory));
+		MatcherAssert.assertThat(
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java:68]: I think for readability, logging test should be a separate test case.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java [hunk 4]
```diff
@@ -42,19 +49,37 @@
...
+			testLoggerResource.getMessages(),
+			hasItem(containsString("The configured or derived JVM heap memory size (50.000mb (52428800 bytes)) is less than its recommended minimum value (128 mb)")));
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java:70]: I think it would be better to also use `jvmHeapSize.toHumanReadableString()` and `JobManagerOptions.MIN_JVM_HEAP_SIZE.toHumanReadableString()` in test so that we do not need to change hard coded values if they change.

Btw, it looks like we forgot to add `toHumanReadableString` for `JobManagerOptions.MIN_JVM_HEAP_SIZE` in `verifyJvmHeapSize()`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java [hunk 5]
```diff
@@ -42,19 +49,37 @@
...
+		jvmHeapMemory = MemorySize.parse("150m");
+
+		conf.set(JobManagerOptions.JVM_HEAP_MEMORY, jvmHeapMemory);
+		conf.set(JobManagerOptions.JOB_STORE_CACHE_SIZE, 200L * 1024L * 1024L);
 
-		JobManagerProcessSpec JobManagerProcessSpec = JobManagerProcessUtils.processSpecFromConfig(conf);
-		assertThat(JobManagerProcessSpec.getJvmHeapMemorySize(), is(jvmHeapSize));
+		jobManagerProcessSpec = JobManagerProcessUtils.processSpecFromConfig(conf);
+		assertThat(jobManagerProcessSpec.getJvmHeapMemorySize(), is(jvmHeapMemory));
+		MatcherAssert.assertThat(
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerProcessUtilsTest.java:79]: Same here, this should be a separate test case.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java [hunk 6]
```diff
@@ -75,12 +75,12 @@ private static JobManagerFlinkMemory createJobManagerFlinkMemory(
...
 			MemorySize offHeapMemory) {
 		verifyJvmHeapSize(jvmHeap);
-		verifyJobStoreCacheSize(config, offHeapMemory);
+		verifyJobStoreCacheSize(config, jvmHeap);
 		return new JobManagerFlinkMemory(jvmHeap, offHeapMemory);
 	}
 
 	private static void verifyJvmHeapSize(MemorySize jvmHeapSize) {
-		if (jvmHeapSize.compareTo(JobManagerOptions.MIN_JVM_HEAP_SIZE) < 1) {
+		if (jvmHeapSize.compareTo(JobManagerOptions.MIN_JVM_HEAP_SIZE) < 0) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java:83]: why is this changed? I think zero is also ok, meaning that `jvmHeapSize` can be exactly the `MIN_JVM_HEAP_SIZE`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java [hunk 7]
```diff
@@ -91,13 +91,13 @@ private static void verifyJvmHeapSize(MemorySize jvmHeapSize) {
 	private static void verifyJobStoreCacheSize(Configuration config, MemorySize jvmHeapSize) {
 		MemorySize jobStoreCacheHeapSize =
 			MemorySize.parse(config.getLong(JobManagerOptions.JOB_STORE_CACHE_SIZE) + "b");
-		if (jvmHeapSize.compareTo(jobStoreCacheHeapSize) < 1) {
+		if (jvmHeapSize.compareTo(jobStoreCacheHeapSize) < 0) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java:94]: same here, why zero?

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java (hunk 6 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java:83]: +1 for using 0; otherwise you can run into the following error which will always be perceived as an oversight:
`The configured or derived JVM heap memory size (128.000mb (134217728 bytes)) is less than its recommended minimum value (128 mb)`

Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java:83]: @zentol I agree with you. @azagrebin The log tells the users the comparation result is "less than".

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java (hunk 7 above)
Comment (SteNicholas) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java:94]: @azagrebin The same reason mentioned above.

**Comment by SteNicholas** [comment]
@azagrebin @zentol I have already add test method `testLogVerifyJvmHeapSizeAndJobStoreCacheSize` for logging test

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java (hunk 6 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/config/memory/jobmanager/JobManagerFlinkMemoryUtils.java:83]: True, sorry for confusion

**Comment by SteNicholas** [comment]
@azagrebin @zentol I have already add test method testLogVerifyJvmHeapSizeAndJobStoreCacheSize for logging test. Please review this change.

**Comment by SteNicholas** [comment]
@azagrebin Yes course.
